Allow an after_connect callback for repo adapters

taylorfinnell opened the pull request:
I needed to be able to access the underlying connection object from Sequel in order to turn on auto commit for a jdbc connection. Since Sequel has an `after_connect` callback I tried to expose it to lotus-model in a reasonable way as well as provide support for the same callback to other repositories for consistency.  Some notes about `after_connect` and its uses: http://sequel.jeremyevans.net/rdoc/files/doc/release_notes/3_11_0_txt.html

My usecase looks like:

```
Lotus::Model.configure do
  adapter type: :sql, uri: 'jdbc:...', after_connect: Proc.new { |connection| connection.auto_commit(true) }

  mapping do
  end
end
```

Please let me know your thoughts about the idea :)

jodosha commented:
@taylorfinnell Hello and thanks for this PR.

Sorry, but I think this approach is too specific. What I would like to see here is a generic set of non-mandatory options to pass down to the adapter. It's then responsible to handle or ignore them.

This opens a world of possibilities larger than callbacks. Eg. if you need to pass a `foo:` option to Sequel, you can do with generic options, but not with the approach above.

Can you please remove the notion of callbacks from the PR and pass down options to the raw adapter (eg `Sequel`)? Once done, I'll be more than happy to merge this. Thank you :+1:

taylorfinnell commented:
@jodosha Thank you for the input! I have made the changes how I believe you want. Please let me know! I am wondering about the MemoryAdapter, currently the options are just ignored. Should I maybe throw an exception about options not being supported for MemoryAdapter, maybe a warning?

jodosha commented on lib/lotus/model/adapters/file_system_adapter.rb:
```diff
@@ -66,7 +66,7 @@ class FileSystemAdapter < MemoryAdapter
         # @api private
         # @since 0.2.0
-        def initialize(mapper, uri)
+        def initialize(mapper, uri, callbacks = {})
```
@taylorfinnell Please change `callbacks` into `options`.

jodosha commented on lib/lotus/model/adapters/abstract.rb:
```diff
@@ -86,9 +86,10 @@ class Abstract
         #
         # @since 0.1.0
-        def initialize(mapper, uri = nil)
+        def initialize(mapper, uri = nil, options = {})
```
@taylorfinnell Can you please document `options` as `@param` in the API doc? Don't forget the other signatures! :smile: Thank you :+1:

taylorfinnell commented:
Updated the variable name, added the docs. Did you mean that I should document all the available options? I suppose the option documentation would go on the adapters themselves. SqlAdapter would note its available options (after_connect, max_connections, etc), MemoryAdapter would note its available options (none)

jodosha commented:
@taylorfinnell What I mean, is to add `@param options [Hash] description here..`

taylorfinnell commented:
@jodosha I think that is all done :)

jodosha commented:
@taylorfinnell It looks good now! Merging :+1:
